tests/integration/test_async_insert_adaptive_busy_timeout/test.py

```diff
@@ -0,0 +1,372 @@
+# the combined duration for fixed timeouts.
+def test_compare_sequential_inserts_durations_for_adaptive_and_fixed_async_timeouts():
+    fixed_tm_table_name = "async_insert_mt_fixed_async_timeout"
+    node.query(
+        "CREATE TABLE {} (a UInt64, b Array(UInt64)) ENGINE=MergeTree() ORDER BY a".format(
+            fixed_tm_table_name
+        )
+    )
+
+    fixed_tm_settings = copy.copy(_query_settings)
+    fixed_tm_settings["async_insert_use_adaptive_busy_timeout"] = 0
+    fixed_tm_settings["async_insert_busy_timeout_ms"] = 200
+
+    fixed_tm_run_duration = timeit.timeit(
+        lambda: _insert_queries_sequentially(
+            fixed_tm_table_name,
+            fixed_tm_settings,
+            iterations=100,
+            max_values_size=1000,
+            array_size_range=[10, 50],
+        ),
+        setup="pass",
+        number=3,
+    )
+
+    node.query("DROP TABLE IF EXISTS {}".format(fixed_tm_table_name))
+
+    logging.debug(
+        "Run duration with fixed asynchronous timeout is {} seconds".format(
+            fixed_tm_run_duration
+        )
+    )
+
+    adaptive_tm_table_name = "async_insert_mt_adaptive_async_timeout"
+    node.query(
+        "CREATE TABLE {} (a UInt64, b Array(UInt64)) ENGINE=MergeTree() ORDER BY a".format(
+            adaptive_tm_table_name
+        )
+    )
+
+    adaptive_tm_settings = copy.copy(_query_settings)
+    adaptive_tm_settings["async_insert_busy_timeout_min_ms"] = 10
+    adaptive_tm_settings["async_insert_busy_timeout_max_ms"] = 1000
+
+    adaptive_tm_run_duration = timeit.timeit(
+        lambda: _insert_queries_sequentially(
+            adaptive_tm_table_name,
+            adaptive_tm_settings,
+            iterations=100,
+            max_values_size=1000,
+            array_size_range=[10, 50],
+        ),
+        setup="pass",
+        number=3,
+    )
+
+    logging.debug(
+        "Run duration with adaptive asynchronous timeout is {} seconds.".format(
+            adaptive_tm_run_duration
+        )
+    )
+
+    node.query("DROP TABLE IF EXISTS {}".format(adaptive_tm_table_name))
+
+    assert adaptive_tm_run_duration <= fixed_tm_run_duration
+
+
+# Ensure that the combined duration of inserts with adaptive timeouts is less than
+# the combined duration for fixed timeouts.
+def test_compare_parallel_inserts_durations_for_adaptive_and_fixed_async_timeouts():
+    fixed_tm_table_name = "async_insert_mt_fixed_async_timeout"
+    node.query(
+        "CREATE TABLE {} (a UInt64, b Array(UInt64)) ENGINE=MergeTree() ORDER BY a".format(
+            fixed_tm_table_name
+        )
+    )
+
+    fixed_tm_settings = copy.copy(_query_settings)
+    fixed_tm_settings["async_insert_use_adaptive_busy_timeout"] = 0
+    fixed_tm_settings["async_insert_busy_timeout_ms"] = 200
+
+    fixed_tm_run_duration = timeit.timeit(
+        lambda: _insert_queries_in_parallel(
+            fixed_tm_table_name,
+            fixed_tm_settings,
+            thread_num=15,
+            tasks=1000,
```

Comment:
@jkartseva can you help me understand why you check with fixed timeout 1000 iterations, but compare it with 100 of adaptive timeout?